### PR TITLE
BUG: merging an Integer EA rasises

### LINKED ISSUE #23020: BUG: merging on an Integer EA raises
```
In [1]: df = pd.DataFrame({'A': pd.Series([1, 2, np.nan], dtype='Int64'), 'B': 1})
```

Merging with an EA included is ok
```
In [2]: pd.merge(df, df, on='B')
Out[2]: 
   A_x  B  A_y
0    1  1    1
1    1  1    2
2    1  1  NaN
3    2  1    1
4    2  1    2
5    2  1  NaN
6  NaN  1    1
7  NaN  1    2
8  NaN  1  NaN

In [3]: pd.merge(df, df, on='B').dtypes
Out[3]: 
A_x    Int64
B      int64
A_y    Int64
dtype: object
```

On the EA column is not Ok
```
In [4]: pd.merge(df, df, on='A').dtypes
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-e77608ca3973> in <module>()
----> 1 pd.merge(df, df, on='A').dtypes

~/pandas/pandas/core/reshape/merge.py in merge(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)
     61                          copy=copy, indicator=indicator,
     62                          validate=validate)
---> 63     return op.get_result()
     64 
     65 

~/pandas/pandas/core/reshape/merge.py in get_result(self)
    562                 self.left, self.right)
    563 
--> 564         join_index, left_indexer, right_indexer = self._get_join_info()
    565 
    566         ldata, rdata = self.left._data, self.right._data

~/pandas/pandas/core/reshape/merge.py in _get_join_info(self)
    771         else:
    772             (left_indexer,
--> 773              right_indexer) = self._get_join_indexers()
    774 
    775             if self.right_index:

~/pandas/pandas/core/reshape/merge.py in _get_join_indexers(self)
    750                                   self.right_join_keys,
    751                                   sort=self.sort,
--> 752                                   how=self.how)
    753 
    754     def _get_join_info(self):

~/pandas/pandas/core/reshape/merge.py in _get_join_indexers(left_keys, right_keys, sort, how, **kwargs)
   1120 
   1121     # get left & right join labels and num. of levels at each location
-> 1122     llab, rlab, shape = map(list, zip(* map(fkeys, left_keys, right_keys)))
   1123 
   1124     # get flat i8 keys from label lists

~/pandas/pandas/core/reshape/merge.py in _factorize_keys(lk, rk, sort)
   1554     elif is_int_or_datetime_dtype(lk) and is_int_or_datetime_dtype(rk):
   1555         klass = libhashtable.Int64Factorizer
-> 1556         lk = ensure_int64(com.values_from_object(lk))
   1557         rk = ensure_int64(com.values_from_object(rk))
   1558     else:   1557         rk = ensure_int64(com.values_from_object(rk))
   1558     else:

~/pandas/pandas/_libs/algos_common_helper.pxi in pandas._libs.algos.ensure_int64()

ValueError: cannot convert float NaN to integer

```

### PR DESCRIPTION
**Pull request by makbigc**
- [ ] closes #23020 
- [ ] 1 test added
- [ ] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you merge master and update

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -1332,6 +1332,12 @@ Sparse
 - Bug in unary inversion operator (``~``) on a ``SparseSeries`` with boolean values. The performance of this has also been improved (:issue:`22835`)
 - Bug in :meth:`SparseArary.unique` not returning the unique values (:issue:`19595`)
 
+IntegerArray
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: just add this onto the big section on IntegerArray which already exists

File: pandas/tests/extension/base/reshaping.py [hunk 2]
```diff
@@ -170,3 +170,11 @@ def test_merge(self, data, na_value):
                  [data[0], data[0], data[1], data[2], na_value],
                  dtype=data.dtype)})
         self.assert_frame_equal(res, exp[['ext', 'int1', 'key', 'int2']])
+
+    def test_merge_on_int_array(self, df_merge_on_int_array):
```
Comment (jreback) [inline pandas/tests/extension/base/reshaping.py]: this needs to be generic, see the test_merge example. it should work for any array type.

File: pandas/tests/extension/conftest.py [hunk 3]
```diff
@@ -104,3 +106,9 @@ def data_for_grouping():
 def box_in_series(request):
     """Whether to box the data in a Series"""
     return request.param
+
+
```
Comment (jreback) [inline pandas/tests/extension/conftest.py]: this is not needed

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by makbigc** [comment]
Please give me some more time. I am working on it.

File: doc/source/whatsnew/v0.24.0.rst [hunk 4]
```diff
@@ -916,6 +916,8 @@ update the ``ExtensionDtype._metadata`` tuple to match the signature of your
 - :meth:`DataFrame.stack` no longer converts to object dtype for DataFrames where each column has the same extension dtype. The output Series will have the same dtype as the columns (:issue:`23077`).
 - :meth:`Series.unstack` and :meth:`DataFrame.unstack` no longer convert extension arrays to object-dtype ndarrays. Each column in the output ``DataFrame`` will now have the same dtype as the input (:issue:`23077`).
 - Bug when grouping :meth:`Dataframe.groupby()` and aggregating on ``ExtensionArray`` it was not returning the actual ``ExtensionArray`` dtype (:issue:`23227`).
+- :meth:`Series.unstack` no longer converts extension arrays to object-dtype ndarrays. The output ``DataFrame`` will now have the same dtype as the input. This changes behavior for Categorical and Sparse data (:issue:`23077`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: i think you picked up an extra whatsnew note here

File: pandas/core/reshape/merge.py [hunk 5]
```diff
@@ -1607,6 +1608,18 @@ def _factorize_keys(lk, rk, sort=True):
 
         lk = ensure_int64(lk.codes)
         rk = ensure_int64(rk)
+    elif (is_period_dtype(lk) and
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: i think you can remove the period_dtype checks as these are now extension arrays

File: pandas/tests/extension/base/reshaping.py [hunk 6]
```diff
@@ -173,6 +173,20 @@ def test_merge(self, data, na_value):
                  dtype=data.dtype)})
         self.assert_frame_equal(res, exp[['ext', 'int1', 'key', 'int2']])
 
+        # GH  23020
```
Comment (jreback) [inline pandas/tests/extension/base/reshaping.py]: can you add as a new test.

this should not be Int64 specific, rather be structured to generically work.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/merge.py (hunk 5 above)
Comment (makbigc) [reply pandas/core/reshape/merge.py]: Removing period_dtype checks would fails test_merge_on_period in test_merge.py.

**Comment by jreback** [comment]
@makbigc can you merge master.

**Comment by makbigc** [comment]
@jreback rebased. Please tell me if anything needs to be done.

**Comment by TomAugspurger** [comment]
There's a linting error: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=4750

```
2018-12-05T15:26:49.5779150Z Check import format using isort
2018-12-05T15:26:55.4954327Z ERROR: /home/vsts/work/1/s/pandas/core/reshape/merge.py Imports are incorrectly sorted.
2018-12-05T15:26:55.5184554Z Check import format using isort DONE
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.rst [hunk 7]
```diff
@@ -1003,6 +1003,7 @@ update the ``ExtensionDtype._metadata`` tuple to match the signature of your
 - :meth:`Series.unstack` and :meth:`DataFrame.unstack` no longer convert extension arrays to object-dtype ndarrays. Each column in the output ``DataFrame`` will now have the same dtype as the input (:issue:`23077`).
 - Bug when grouping :meth:`Dataframe.groupby()` and aggregating on ``ExtensionArray`` it was not returning the actual ``ExtensionArray`` dtype (:issue:`23227`).
 - A default repr for :class:`ExtensionArray` is now provided (:issue:`23601`).
+- Bug in :func:`pandas.merge` when merging on an Integer extension array (:issue:`23020`)
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply doc/source/whatsnew/v0.24.0.rst]: This fixes merging on any extension array right? That should be noted in the whatsnew.

File: pandas/tests/extension/base/reshaping.py (hunk 6 above)
Comment (TomAugspurger) [reply pandas/tests/extension/base/reshaping.py]: The test will need to be updated to merge on the extension array data's column.

**Comment by makbigc** [comment]
The test is rewritten. So the EA is the merging key, not like the previous test in which integer array is the merging key for every cases.
The old test is placed as a special case in the tests/reshape/merge/test_merge.py.
Would you give me some hints to pass azure test? It seems the new test doesn't work with python 2.7.15.
Thanks.

File: pandas/core/arrays/period.py [hunk 8]
```diff
@@ -814,6 +814,19 @@ def _check_timedeltalike_freq_compat(self, other):
     def _values_for_argsort(self):
         return self._data
 
+    def _values_for_factorize(self):
```
Comment (TomAugspurger) [inline pandas/core/arrays/period.py]: This method already exists on master. Can you try merging master, and keeping the implementation from master?

File: pandas/core/arrays/sparse.py [hunk 9]
```diff
@@ -942,7 +942,7 @@ def unique(self):
 
     def _values_for_factorize(self):
         # Still override this for hash_pandas_object
-        return np.asarray(self), self.fill_value
+        return np.asarray(self, object), self.fill_value
```
Comment (TomAugspurger) [inline pandas/core/arrays/sparse.py]: I don't think we want object-dtype here. Why did you need it?

File: pandas/core/reshape/merge.py [hunk 10]
```diff
@@ -1533,6 +1534,10 @@ def _factorize_keys(lk, rk, sort=True):
 
         lk = ensure_int64(lk.codes)
         rk = ensure_int64(rk)
+    elif is_extension_array_dtype(lk) and is_extension_array_dtype(rk):
+        klass = libhashtable.Factorizer
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/reshape/merge.py]: @makbigc did you try @jreback's suggesting of factorizing the EAs *before* getting into the main if block?

File: pandas/tests/extension/base/reshaping.py [hunk 11]
```diff
@@ -173,6 +173,21 @@ def test_merge(self, data, na_value):
...
+        res = pd.merge(df1, df1, on='key')
+
+        exp = pd.DataFrame(
+            {'key': data[:3],
+             'ext_x': [1, 2, 3],
+             'ext_y': [1, 2, 3]})
+
+        self.assert_frame_equal(res, exp[['ext_x', 'key', 'ext_y']],
+                                check_dtype=True)
+
```
Comment (TomAugspurger) [inline pandas/tests/extension/base/reshaping.py]: It'd be good to have tests that aren't a one-to-one merge. Something like

```python
a, b = data[:2]

key1 = type(data)._from_sequence([a])
key2 = type(data)._from_sequence([a, b])
```


And put those in a DataFrame and do an inner / outer join. LMK if you want help with that.

**Comment by TomAugspurger** [comment]
@makbigc I'm going to push a few changes here if that's OK.

**Comment by TomAugspurger** [comment]
I started to write a test that would include NA values in the merge key, but I think the expected behavior of that is up in the air right now: https://github.com/pandas-dev/pandas/issues/22491.

File: pandas/core/reshape/merge.py [hunk 12]
```diff
@@ -1593,6 +1594,15 @@ def _factorize_keys(lk, rk, sort=True):
         lk = lk.values
         rk = rk.values
 
+    elif (is_extension_array_dtype(lk.dtype) and
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: maybe should do this after the categorical check?

File: pandas/tests/reshape/merge/test_merge.py [hunk 13]
```diff
@@ -1326,6 +1326,16 @@ def test_merging_with_bool_or_int_cateorical_column(self, category_column,
             CDT(categories, ordered=ordered))
         assert_frame_equal(expected, result)
 
+    def test_merge_on_int_array(self):
+        # GH 23020
+        df = pd.DataFrame({'A': pd.Series([1, 2, np.nan], dtype='Int64'),
+                           'B': 1})
+        result = pd.merge(df, df, on='A')
```
Comment (jreback) [inline pandas/tests/reshape/merge/test_merge.py]: I think can remove this?

File: pandas/core/reshape/merge.py (hunk 12 above)
Comment (TomAugspurger) [reply pandas/core/reshape/merge.py]: The categorical check is already in the main `if` block.

I'm going to move that up to the same if / elif as the datetimetz, and the new extension dtype check.

File: pandas/tests/reshape/merge/test_merge.py (hunk 13 above)
Comment (TomAugspurger) [reply pandas/tests/reshape/merge/test_merge.py]: I guess this does test merging on an NA key, which the generic test doesn't. So maybe leave it?

**Comment by TomAugspurger** [comment]
Updated the coercion of lk / rk to do

1. all the pre-processing on the values (datetimetz, categorical, extension)
2. Hashtable / dtype validate

in two distinct blocks.

**Comment by makbigc** [comment]
@TomAugspurger Thanks for your reply and help.

**Review by jreback — APPROVED** [review]
